src/pip/_internal/resolution/resolvelib/factory.py

```diff
@@ -205,6 +205,31 @@ def _iter_found_candidates(
 
         return six.itervalues(candidates)
 
+    def _filter_explicit_candidates(
+        self,
+        explicit_candidates,  # type: Set[Candidate]
+        requirements,  # type: Sequence[Requirement]
+    ):
+        # type: (...) -> Iterable[Candidate]
+        for c in explicit_candidates:
+            if not all(req.is_satisfied_by(c) for req in requirements):
+                continue
+            dist = self._installed_dists.get(c.name)
+            if dist:
+                extras = getattr(c, "extras", frozenset())
+                template = c.get_install_requirement()
+                # Does not work because an AlreadyInstalledCandidate
+                # does not satisfy the ExplicitRequirement :-(
+                if template:
+                    # TODO: Can template be None here?
```

Comment:
I'd put an assert and wait for the beta. :P

That way, we'll know pretty quickly during the beta if it is, and if not, that assertion doesn't distract from the change.

Reply:
Given that I can't see a way to get around the `AlreadyInstalledCandidate` vs `ExplicitRequirement` issue, this is at best a minor point (but I would agree with asserting otherwise).

Question: How bad would it be if an `ExplicitRequirement` were changed to be saisfied, not only by the candidate it was built from, but *also* by an `AlreadyInstalled Candidate` that had the same name and version? I think that yesterday I assumed this wasn't acceptable, but on reflection (and after a night's sleep) I wonder if it would be OK. It might add some annoying coupling between `requirements.py` and `candidates.py`, but we can refactor that later...

I'll try that later.

Reply:
Looks like that broke a lot of stuff 🙁